### DISCUSSION TITLE
Permission to view commit info, without contents permissions

### DISCUSSION THREAD

**Discussion by LasseRafn** [community/community#62476] [Apps, API and Webhooks]
### Select Topic Area

Question

### Body

I'd suggest opening up the [commit endpoints](https://docs.github.com/en/rest/commits/commits?apiVersion=2022-11-28) to be accessible through a separate permission than "contents" as it permits reading code.

If I want to use the GitHub API and permit companies doing anything that connects with their commits, I **HAVE TO** request full code access. Thats way too broad permissions, just to get basic commit info (author, SHA, date, message, ...).

Currently it is possible to get commit info through [Pull Requests permissions](https://docs.github.com/en/rest/pulls/pulls?apiVersion=2022-11-28#list-commits-on-a-pull-request), so I would argue that generic commit endpoints should not require contents permissions either. Seems like an oversight that you can view PR commits but not commits, without code access. There is no data in this endpoint that I can't see in the other (Contents-permission) endpoint, except when switching response type to diff/patch; which would of course be restricted to "Contents" permission, even with this change.

Either through a new "Commits" permission (ideally), or under one of the other less 'risky' permissions.

This would help developers build apps to GitHub that focuses on code privacy, as well as make GitHub organizations more comfortable installing apps, without worrying that their intellectual property can be accessed, for apps that never need to have such access.

If I could access commit history without being able to access patch/diff + code, that would solve the issue. 

**Examples of apps that could benefit:**

- Some sort of commit history aggregator across teams
- Test coverage trackers (in most cases, they don't need access to code, just git history)
- Jira GitHub app (yeah it has read/write access to code, despite only needing to link commits to your tickets etc.!)
- Surely many many more

**Comment by ezzeddinp** [comment]
I can provide some insights and suggestions:

1. Feature Request: You can submit a feature request to GitHub directly through their official channels. Visit the GitHub Community Forum or GitHub's official feedback page to share your suggestions about separating commit-related endpoints from "contents" permission. GitHub's development team actively monitors these channels and considers user feedback for future improvements.

2. GitHub GraphQL API: Instead of using the traditional REST API, consider exploring GitHub's GraphQL API. GraphQL allows you to request only the specific fields and data you need, providing more fine-grained control over the information you retrieve. This could potentially help you obtain commit information without granting broader access to code contents.

3. OAuth Scopes: When requesting OAuth access to a user's account for your application, you can specify the necessary scopes for the permissions your app requires. By carefully selecting the scopes, you can limit the access to only the information necessary for your app's functionality.

4. Token Scopes: If you are using personal access tokens (PATs) for your application, GitHub allows you to set token scopes to limit the permissions associated with the token. Again, you can carefully choose the required scopes to access commit information without granting broader code access.

5. Third-Party Apps: If you are developing an application for GitHub, you can clarify the specific permissions your app requires to the potential users. By being transparent about your app's intentions and the data it accesses, you can build trust with organizations and users.

Keep in mind that any changes to GitHub's API or permissions system would be subject to GitHub's own considerations, security concerns, and developer community feedback. By sharing your suggestions and use cases through appropriate channels, you can contribute to shaping the future direction of the GitHub platform and its APIs.

Hopefully this answer your question !!
like to answer if you has any question by follow and chat me

**Reply by LasseRafn** [reply]
Thanks @ezzeddinp ! I can't seem to get either of the options to work, as commits are guarded behind 'contents' :(.. also for GraphQL.

How would I go about submitting a feature request? I thought this was the official channel for that, would you happen to know where? 🙏🏻 

Regardless, thanks for your time and help

**Reply by ezzeddinp** [reply]
To submit a feature request for GraphQL, you can follow these general steps:

- Identify the official channel for submitting feature requests. This can vary depending on the platform or service you are using. For example:
Salesforce: You can submit feature requests through the Salesforce Ideas platform or contact Salesforce support for guidance developer.salesforce.com.
GitHub: You can submit feature requests through the GitHub Community Forum or open an issue in the relevant GitHub repository docs.github.com.
Azure API Management: You can submit feature requests through the Azure Feedback Forum or contact Azure support for assistance learn.microsoft.com.


Before submitting your feature request, it's helpful to gather relevant information:

> Clearly describe the feature you are requesting and its expected benefits.

> Provide use cases or examples that highlight the need for the feature.

> Check if the feature has already been requested or discussed in the platform's documentation, forums, or issue tracker. If it has, you can add your comments or upvote the existing request to show support.

> Include any relevant technical details or specifications that may be helpful for understanding and implementing the feature.
Submit your feature request through the appropriate channel. Follow any guidelines or templates provided by the platform to ensure your request is properly documented and considered.

Remember that feature requests are evaluated by the platform or service provider based on various factors such as user demand, technical feasibility, and alignment with the platform's goals. It's important to provide a clear and compelling case for the requested feature to increase the chances of it being considered and implemented.

If you are unsure about the specific process for submitting feature requests for a particular platform or service, it's recommended to consult the official documentation or reach out to the platform's support channels for guidance.


Submit a feature request for GraphQL by following the general steps provided...


Sure! Here's an example of how you can submit a feature request for GraphQL:

- Identify the official channel: Determine the platform or service you are using GraphQL with and find the official channel for feature requests. Let's say you are using GitHub as an example.

- Research existing requests: Before submitting your request, it's a good idea to check if someone else has already requested a similar feature. This can be done by searching the GitHub repository's issues or checking the GitHub Community Forum.

- Prepare your feature request:

> Clearly describe the feature: Provide a concise and descriptive title for your request. In the body, explain the feature in detail. Include examples or use cases to demonstrate its usefulness.

> Explain the motivation: Why is this feature important? Describe the problem or limitation that the feature would address and how it would benefit users.

> Provide any technical details: If applicable, include any specifications or technical information that can help with implementing the feature. This could include details about the GraphQL schema or specific API endpoints that need to be modified or added.

> Submit your feature request:
GitHub: If you are using GraphQL with GitHub, you can submit a feature request by opening an issue in the relevant repository. Ensure that you are in the correct repository where the improvement is needed, and navigate to the "Issues" tab. Click on the "New Issue" button and fill out the issue template with the details of your feature request.

Remember to provide a clear and concise title, followed by a detailed description of the feature. You can use the GitHub formatting options to format your request with headings, bullet points, and code blocks if necessary. Once you've filled out all the required information, click on "Submit new issue" to officially submit your feature request.

Depending on the platform or service, the process for submitting feature requests may differ, and there may be specific guidelines or templates to follow. It's always a good idea to review the documentation or contact support for the platform you're using to ensure you're following the correct process.

By following these general steps, you should be able to effectively submit a feature request for GraphQL or any other platform or service you are using.

I hope this helps! Let me know if you have any further questions.

**Comment by ghostinhershell** [comment]
Hey @LasseRafn, thanks so much for reaching out in our Community! I've added the 'Enhancement' label to this discussion so that it can be tagged as a feature request. Let us know if you have any additional questions. Cheers!

**Comment by LasseRafn** [comment]
@ghostinhershell thanks for this! Hope this can be taken up at some point, would enable GitHub-app developers to secure against issues such as these when code access is not needed, while still having access to necessary data:

- https://github.blog/news-insights/company-news/security-alert-stolen-oauth-user-tokens/
- https://www.reuters.com/article/ctech-us-usa-cyber-codecov-idCAKBN2C62XI-OCATC

**Comment by hkdobrev** [comment]
I was able to read enough of what I needed by reading PR timeline items filtered by item type for just commits. These are accessible via GraphQL without the contents permission, but just PR read permission.

https://docs.github.com/en/graphql/reference/objects#pullrequesttimelineconnection

**Comment by rwellington** [comment]
Hi @ghostinhershell 👋 I have the same use case as @LasseRafn. It'd be great not have to request access to the code-content in order to access commit metadata from the fine grain access controls. It'd be a big win for the GitHub community and unlock a lot of possibilities (while closing security risks for companies). What might we do to get this implemented? Anything we can help with?

**Comment by brunohaid** [comment]
Same here, we run deliberately simple GIT worklflows (often simple commit & push, no PRs etc), but still want programmatic access to commit metadata without having to deploy keys around the world that would give any compromised machine full access to our entire codebase.

Any chance to get a simple `Contents:metadata` read for that, similar to what the Events and PR APIs are exposing?
